dotnet/src/Connectors/Connectors.Memory.Redis/RedisMemoryStore.cs

```diff
@@ -0,0 +1,269 @@
+            {
+                yield return result;
+            }
+            else
+            {
+                yield break;
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<string> UpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancellationToken = default)
+    {
+        record.Key = record.Metadata.Id;
+
+        await this._database.HashSetAsync(GetRedisKey(collectionName, record.Key), new[] {
+            new HashEntry("key", record.Key),
+            new HashEntry("metadata", record.GetSerializedMetadata()),
+            new HashEntry("embedding", MemoryMarshal.Cast<float, byte>(record.Embedding.AsReadOnlySpan()).ToArray()),
+            new HashEntry("timestamp", ToTimestampLong(record.Timestamp))
+        }, flags: CommandFlags.None).ConfigureAwait(false);
+
+        return record.Key;
+    }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<string> UpsertBatchAsync(string collectionName, IEnumerable<MemoryRecord> records, [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        foreach (var record in records)
+        {
+            yield return await this.UpsertAsync(collectionName, record, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task RemoveAsync(string collectionName, string key, CancellationToken cancellationToken = default)
+    {
+        await this._database.KeyDeleteAsync(GetRedisKey(collectionName, key), flags: CommandFlags.None).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc/>
+    public async Task RemoveBatchAsync(string collectionName, IEnumerable<string> keys, CancellationToken cancellationToken = default)
+    {
+        await this._database.KeyDeleteAsync(keys.Select(key => GetRedisKey(collectionName, key)).ToArray(), flags: CommandFlags.None).ConfigureAwait(false);
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<(MemoryRecord, double)> GetNearestMatchesAsync(
+        string collectionName,
+        Embedding<float> embedding,
+        int limit,
+        double minRelevanceScore = 0,
+        bool withEmbeddings = false,
+        [EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        if (limit <= 0)
+        {
+            yield break;
+        }
+
+        var query = new Query($"*=>[KNN {limit} @embedding $embedding AS vector_score]")
+                    .AddParam("embedding", MemoryMarshal.Cast<float, byte>(embedding.AsReadOnlySpan()).ToArray())
+                    .SetSortBy("vector_score")
+                    .ReturnFields("key", "metadata", "embedding", "timestamp", "vector_score")
+                    .Limit(0, limit)
+                    .Dialect(2);
+
+        var results = await this._ft.SearchAsync(collectionName, query).ConfigureAwait(false);
+
+        foreach (var document in results.Documents)
+        {
+            double score = 1 - (double)document["vector_score"];
```

Comment:
Are we sure this casting will always be successful? Should we wrap it with try/catch and throw appropriate exception if something went wrong?

Reply:
This value is calculated by RediSearch, and RedisValue is automatically converted for numeric strings. Although the probability of returning a non-numeric value is extremely small (RediSearch has changed some rules), a more appropriate exception would be better.